irc_send.c

```diff
@@ -293,15 +287,26 @@ void irc_send_who(irc_t *irc, GSList *l, const char *channel)
 	gboolean is_channel = strchr(CTYPES, channel[0]) != NULL;
 
 	while (l) {
-		irc_user_t *iu = l->data;
+		irc_user_t *iu;
+
+		/* Null terminated string with three chars, respectively:
+		 * { <H|G>, <@|%|+|\0>, \0 } */
+		char status_prefix[3] = {0};
+
 		if (is_channel) {
-			iu = ((irc_channel_user_t *) iu)->iu;
+			irc_channel_user_t *icu = l->data;
+			status_prefix[1] = irc_channel_user_get_prefix(icu);
+			iu = icu->iu;
+		} else {
+			iu = l->data;
 		}
-		/* TODO(wilmer): Restore away/channel information here */
-		irc_send_num(irc, 352, "%s %s %s %s %s %c :0 %s",
+
+		/* rfc1459 doesn't mention this: G means gone, H means here */
+		status_prefix[0] = iu->flags & IRC_USER_AWAY ? 'G' : 'H';
```

Comment:
That's actually reformatted from your old code, `iu->flags & IRC_USER_AWAY ? 'G' : 'H'`, a couple of lines below. I just added the comment (because they confused me) and removed the TODO (which wasn't valid anymore)

re: moving up, good idea.